### PR TITLE
refactor: Refactor ML pipeline into modular `riichienv-ml` package and modernize core observation encoding

### PR DESCRIPTION
**Pull request by smly**
The `demos/ml_baseline/` directory grew organically and mixed prototyping code. This refactoring establishes a clear package boundary (`riichienv-ml`) with well-defined sub-modules, making it easier to iterate on model architectures and training pipelines independently from the core environment.

Should split this up into more PRs. This PR is too long to review.😇

## Summary

- Relocate the experimental `demos/ml_baseline/` code into a dedicated `riichienv-ml/` package with a clean modular structure (datasets, models, trainers, features, configs)
- Move UI-related demos (`replay_debug`, `replay_demo`, `replay_penalty_display`) to `riichienv-ui/demos/`
- Modernize 3-player observation encoding and helpers for consistency with the 4-player pipeline
- Add sequence feature encoding and replay event progression caching in `riichienv-core` to support transformer-based models

### `riichienv-ml/` (new package)
- **Datasets**: Modular dataset classes for GRP, MJAI logs, PPO rollouts, and hand prediction
- **Models**: Shared CNN backbone, transformer encoder, actor-critic, Q-network, GRP model, and hand prediction model with cross-attention policy head
- **Trainers**: BC (behavioral cloning from logs and from model), CQL, GRP, PPO, and hand prediction trainers
- **Features**: Pluggable feature extractors (`feat_v1`/`v2`/`v3`) and sequence feature encoding
- **Configs**: YAML configurations for 3p and 4p training modes (BC, CQL, GRP, PPO, hand prediction)
- **Scripts**: Standalone entry points for training and inference

### `riichienv-core/` (Rust)
- Add `sequence_features` module for sparse/dense/progression encoding used by transformer models
- Implement Python-side replay event progression caching for O(1) access
- Refine MJAI event handling for melds, kita, and hule in replay parser
- Update sanma (3p) observation encode shapes to 27 tiles
- Improve readability of observation and state event processing

### CODE REVIEW

File: riichienv-ml/src/riichienv_ml/models/grp_model.py [hunk 1]
```diff
@@ -24,76 +24,68 @@ def forward(self, x):
...
-            row["p1_delta_score"],
-            row["p2_delta_score"],
-            row["p3_delta_score"],
-        ])
+    def calc_pts_reward(self, row: dict, player_idx: int) -> np.ndarray:
+        n = self.n_players
+        scores = np.array([row[f"p{i}_init_score"] for i in range(n)]
+                          + [row[f"p{i}_end_score"] for i in range(n)])
+        delta_scores = np.array([row[f"p{i}_delta_score"] for i in range(n)])
         scores = scores / 25000.0
```
[the remark this answers is not in the document]
Comment (smly) [reply riichienv-ml/src/riichienv_ml/models/grp_model.py]: Addressed together with https://github.com/smly/RiichiEnv/pull/114#discussion_r2867103701. The `_score_norm` attribute is used consistently in both methods.

File: riichienv-ml/src/riichienv_ml/datasets/grp_dataset.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        files = sorted(glob.glob(self.data_glob, recursive=True))
+        return files
+
+    def _encode_features(self, grp_features: dict, player_idx: int) -> torch.Tensor:
+        n = self.n_players
+
+        scores = np.array(
+            [grp_features[f"p{i}_init_score"] / 25000.0 for i in range(n)]
+            + [grp_features[f"p{i}_end_score"] / 25000.0 for i in range(n)]
+            + [grp_features[f"p{i}_delta_score"] / 12000.0 for i in range(n)],
```
[the remark this answers is not in the document]
Comment (smly) [reply riichienv-ml/src/riichienv_ml/datasets/grp_dataset.py]: Fixed. `_encode_features` now computes `score_norm = 35000.0 if n == 3 else 25000.0` locally and uses it for all score normalization.

File: riichienv-ml/src/riichienv_ml/datasets/mjai_logs.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        self.n_players = n_players
+        self.replay_rule = replay_rule
+        self.encoder = encoder
+
+    def _get_files(self):
+        if isinstance(self.data_sources, list):
+            return self.data_sources
+        elif isinstance(self.data_sources, str):
+            return glob.glob(self.data_sources)
+        return []
```
[the remark this answers is not in the document]
Comment (smly) [reply riichienv-ml/src/riichienv_ml/datasets/mjai_logs.py]: Good catch. Added `recursive=True` to the `glob.glob()` call in the string branch of `_get_files()`, matching the convention used in the rest of the pipeline (e.g., `GrpReplayDataset._get_files()` and `Trainer.__init__`).

File: riichienv-ml/src/riichienv_ml/trainers/grp.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        optimizer = optim.Adam(model.parameters(), lr=self.lr)
+        t_max = n_epochs * self._estimate_steps_per_epoch()
+        logger.info(f"CosineAnnealingLR: T_max={t_max} (files={self.n_train_files}, "
+              f"samples_per_file={self.samples_per_file}, batch_size={self.batch_size})")
+        scheduler = CosineAnnealingLR(optimizer, T_max=t_max, eta_min=self.lr_eta_min)
+        criterion = nn.CrossEntropyLoss()
+        for epoch in range(n_epochs):
+            train_loss, train_acc = self._train_epoch(epoch, model, optimizer, scheduler, criterion)
+            val_loss, val_acc = self._val_epoch(epoch, model, criterion)
+            torch.save(model.state_dict(), output_path)
```
[the remark this answers is not in the document]
Comment (smly) [reply riichienv-ml/src/riichienv_ml/trainers/grp.py]: Fixed. The validation step is now guarded with `hasattr(self, "val_dataloader")`. When no validation data is provided, training proceeds without validation metrics and only `train/loss`, `train/acc`, and `lr` are logged to W&B.

File: riichienv-ml/src/riichienv_ml/trainers/grp.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+                n_players=n_players,
+                replay_rule=replay_rule,
+                is_train=True,
+            )
+            self.train_dataloader = DataLoader(
+                self.train_dataset,
+                batch_size=batch_size,
+                num_workers=num_workers,
+                pin_memory=True,
+            )
```
[the remark this answers is not in the document]
Comment (smly) [reply riichienv-ml/src/riichienv_ml/trainers/grp.py]: Added an early `ValueError` at the top of `train()` that fires when `train_dataloader` was not created (i.e., `data_glob` was empty). This gives a clear error message: `"data_glob is required for training (got empty string)"`.

File: riichienv-ml/src/riichienv_ml/utils.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+            mode="w", suffix=".diff", prefix="git_diff_", delete=False,
+        ) as f:
+            f.write(diff)
+            diff_path = f.name
+        try:
+            art_diff.add_file(diff_path, name="git_diff.diff")
+            run.log_artifact(art_diff)
+            logger.info("Uploaded code-diff artifact (uncommitted changes)")
+        finally:
+            os.unlink(diff_path)
```
[the remark this answers is not in the document]
Comment (smly) [reply riichienv-ml/src/riichienv_ml/utils.py]: Added a `logger.warning()` before the upload that explicitly alerts the user: *"Uploading uncommitted git diff to W&B. Ensure no secrets/credentials are in your working tree."* This makes the behavior transparent without breaking existing workflows. A full opt-in flag can be added if this becomes a concern in practice.
